drivers/accel/adxl367/adxl367.c

```diff
@@ -0,0 +1,1310 @@
+ *
+ * @param dev      - The device structure.
+ * @param odr      - Output Data Rate option.
+ *					 Accepted values: 	ADXL367_ODR_12P5HZ,
+ *										ADXL367_ODR_25HZ,
+ *										ADXL367_ODR_50HZ,
+ *										ADXL367_ODR_100HZ,
+ *										ADXL367_ODR_200HZ,
+ *										ADXL367_ODR_400HZ
+ *
+ * @return 0 in case of success, negative error code otherwise.
+*******************************************************************************/
+int32_t adxl367_set_output_rate(struct adxl367_dev *dev,
+				enum adxl367_odr odr)
+{
+	int32_t ret;
+
+	ret = adxl367_reg_write_msk(dev,
+				    ADXL367_REG_FILTER_CTL,
+				    ADXL367_FILTER_CTL_ODR(odr),
+				    ADXL367_FILTER_CTL_ODR(0x7));
+	if (ret)
+		return ret;
+
+	dev->odr = odr;
+
+	return 0;
+}
+
+/***************************************************************************//**
+ * @brief Sets offset for each axis.
+ *
+ * @param dev      - The device structure.
+ * @param x_offset - X axis offset.
+ * @param y_offset - Y axis offset.
+ * @param z_offset - Z axis offset.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+*******************************************************************************/
+int32_t adxl367_set_offset(struct adxl367_dev *dev, uint16_t x_offset,
+			   uint16_t y_offset, uint16_t z_offset)
+{
+	int32_t ret;
+
+	ret = adxl367_reg_write_msk(dev, ADXL367_REG_X_OFFSET,
+				    x_offset & ADXL367_XYZ_AXIS_OFFSET_MASK,
+				    ADXL367_XYZ_AXIS_OFFSET_MASK);
+	if (ret)
+		return ret;
+	ret = adxl367_reg_write_msk(dev, ADXL367_REG_Y_OFFSET,
+				    y_offset & ADXL367_XYZ_AXIS_OFFSET_MASK,
+				    ADXL367_XYZ_AXIS_OFFSET_MASK);
+	if (ret)
+		return ret;
+	return adxl367_reg_write_msk(dev, ADXL367_REG_Z_OFFSET,
+				     z_offset & ADXL367_XYZ_AXIS_OFFSET_MASK,
+				     ADXL367_XYZ_AXIS_OFFSET_MASK);
+}
+
+/***************************************************************************//**
+ * @brief Reads the 3-axis raw data from the accelerometer.
+ *
+ * @param dev - The device structure.
+ * @param x   - Stores the X-axis data(as two's complement).
+ * @param y   - Stores the Y-axis data(as two's complement).
+ * @param z   - Stores the Z-axis data(as two's complement).
+ *
+ * @return 0 in case of success, negative error code otherwise.
+*******************************************************************************/
+int32_t adxl367_get_raw_xyz(struct adxl367_dev *dev,
+			    int16_t* x,
+			    int16_t* y,
+			    int16_t* z)
+{
+	int32_t ret;
+	uint8_t xyz_values[6] = {0, 0, 0, 0, 0, 0};
+	uint8_t reg_data, nready = 1U;
+
+	while (nready) {
+		ret = adxl367_get_register_value(dev, &reg_data, ADXL367_REG_STATUS, 1);
+		if (ret)
+			return -1;
+		if (reg_data & ADXL367_STATUS_DATA_RDY)
+			nready = 0U;
+	}
+
+	ret = adxl367_get_register_value(dev,
+					 xyz_values,
+					 ADXL367_REG_XDATA_H,
+					 6);
+
+	// result is 14 bits long, ignore last 2 bits from low byte
+	*x = ((int16_t)xyz_values[0] << 6) + (xyz_values[1] >> 2);
```

Comment:
wondering if all this computing can't be done with GENMASK + FIELD_PREP/FIELD_GET? Just a question

Reply:
I think the code is easier to read this way.